### PR TITLE
feat: added first draft of replacement between delimiter

### PR DESCRIPTION
**Pull request by Yingrjimsch**
Implementation of [Issue 5616](https://github.com/kubernetes-sigs/kustomize/issues/5616)
#5616

### CODE REVIEW

**Comment by Yingrjimsch** [comment]
Edit: I decided to add a fulltext replacement as well because it was very similar to the start and enddelimiter. Should not be too much to review and is pretty straightforward.

**Comment by olegmayko** [comment]
hi, any ETA on the merge of this PR?

**Comment by Yingrjimsch** [comment]
> hi, any ETA on the merge of this PR? 

I tried via Slack and Assigning to find someone but until now I got sadly no reaction...

**Review by harshsingh32 — COMMENTED** [review]
lgtm

**Comment by newtondev** [comment]
This is a great feature and would love to start using it. There any reviewers that can approve?

File: api/filters/replacement/replacement.go [hunk 1]
```diff
@@ -93,7 +98,24 @@ func getRefinedValue(options *types.FieldOptions, rn *yaml.RNode) (*yaml.RNode,
...
+	} else {
+		mapper := func(s string) string {
+			s = strings.ReplaceAll(s, options.Delimiter, "")
+			s = strings.ReplaceAll(s, options.EndDelimiter, "")
+			return s
+		}
+		if options.Delimiter == "" {
+			return nil, fmt.Errorf("delimiter needs to be set if enddelimiter is set")
+		}
+		re := regexp.MustCompile(regexp.QuoteMeta(options.Delimiter) + `(.*?)` + regexp.QuoteMeta(options.EndDelimiter))
```
Comment (varshaprasad96) [inline api/filters/replacement/replacement.go:113]: This seems to be a bit fragile:
- We assume that delimiter and end delimiter are in pairs, but this may not be the case always. 
- Also what happens when delimiters are not in expected order, or there are nested delimiters. 

If not for nested delimiters, for the other cases, can we do any kind of pre-processing?

File: api/filters/replacement/replacement.go [hunk 2]
```diff
@@ -247,3 +267,32 @@ func setFieldValue(options *types.FieldOptions, targetField *yaml.RNode, value *
...
+		tv = append([]string{source}, tv...)
+	case index >= len(tv): // suffix
+		tv = append(tv, source)
+	default: // replace an element
+		tv[index] = source
+	}
+	return strings.Join(tv, delimiter)
+}
+
+func getByRegex(regex string, target string, source string, index int) string {
```
Comment (varshaprasad96) [inline api/filters/replacement/replacement.go:285]: Since both these are fragile string manipulation methods, can we have separate tests to verify these and ensure we are capturing the edge cases?

File: api/filters/replacement/replacement.go (hunk 1 above)
Comment (Yingrjimsch) [reply api/filters/replacement/replacement.go:113]: @varshaprasad96 thank you for the feedback. 

- The option is that `Delimiter` comes alone, in this case it stays like it is today, the `EndDelimiter` should not be used without a `Delimiter` if so we return an error and if they are used in combination, the value in between them is loaded. If I'm not mistaken all cases should be handled.
- The delimiters should always be in expected order because their key (delimiter or enddelimiter) allow us to define the order.
- To make it less fragile we could do a preprocessing, if the `EndDelimiter` is set and the `Delimiter` is empty they swap values. Do you have another suggestion for preprocessing?

File: api/filters/replacement/replacement.go (hunk 2 above)
Comment (Yingrjimsch) [reply api/filters/replacement/replacement.go:285]: @varshaprasad96 I've written some edge case tests but if you want I can add some tests that handle these functions in specific

File: api/filters/replacement/replacement.go (hunk 1 above)
Comment (Yingrjimsch) [reply api/filters/replacement/replacement.go:120]: @varshaprasad96 I will look into this this week and give you a feedback. If I see a possibility to ensure the order and the amount of delimiters I would implement it so you can review it.

Comment (Yingrjimsch) [reply api/filters/replacement/replacement.go:120]: Hi @varshaprasad96 I've looked into it and wanted to ask if you could provide a nested example for me. As I have seen the delimiters are used in `FieldOptions` https://github.com/kubernetes-sigs/kustomize/blob/master/api/types/replacement.go#L67 which contains of primitive datatypes variables, therefore I don't see how multiple delimiters or nesting could accure. The replacement consists of one `SourceSelector` and multiple `TargetSelector` everyone of these types contain exactly one `FieldOptions` type. 

If I am missing something just tell me and I will look into it.

**Comment by koba1t** [comment]
/hold 
Sorry, but I add a comment to the feature request issue.
